### PR TITLE
TP05

### PR DESCRIPTION
**Pull request by Cristiana959**
### Pull Request Overview

This pull request adds TP04.

### Testing Strategy

This pull request was tested by...
N//A

### TODO or Help Wanted

N/A

### Build

- [ ] Ran `npm build`.

### Author

Signed-off-by: Cristiana Andrei cristiana.andrei@wyliodrin.com

### CODE REVIEW

File: docs/tp/04.md [hunk 1]
```diff
@@ -0,0 +1,60 @@
+---
+sidebar_position: 4
+description: Le but de ce TP est d'apprendre à utiliser des descripterus de fichers et dup2
```
Comment (alexandruradovici) [inline docs/tp/04.md:3]:
```suggestion
description: Utilisation des descripterus de fichers et dup2
```

File: docs/tp/04.md [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+
+Vous pouvez donc voir qu'il existe deux flux de sortie, `stdout` et `stderr`, et un flux d'entrée, `stdin`. Étant donné que les messages d'erreur et la sortie normale ont chacun leur propre conduit pour les acheminer vers la fenêtre du terminal, ils peuvent être traités indépendamment les uns des autres.
+
+Les flux sous Linux, comme presque tout le reste, sont traités comme s'ils étaient des **fichiers**. Vous pouvez lire du texte à partir d'un fichier et vous pouvez écrire du texte dans un fichier. Ces deux actions impliquent un flux de données. Ainsi, le concept de traitement d'un flux de données en tant que fichier n'est pas si exagéré.
+
+:::note
+Chaque fichier associé à un processus se voit attribuer un **numéro unique** pour l'identifier. C'est ce qu'on appelle le **descripteur de fichier**. Chaque fois qu'une action doit être effectuée sur un fichier, le descripteur de fichier est utilisé pour identifier le fichier.
```
Comment (alexandruradovici) [inline docs/tp/04.md:24]:
```suggestion
Chaque instance fichier ouverte associé à un processus se voit attribuer un **numéro unique** pour l'identifier. C'est ce qu'on appelle le **descripteur de fichier**. Chaque fois qu'une action doit être effectuée sur un fichier, le descripteur de fichier est utilisé pour identifier le fichier.
```
 Please verify this translation

File: docs/tp/04.md [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+
+### dup2
+L'appel système `dup2()` est similaire à `dup()` mais la différence fondamentale entre eux est qu'au lieu d'utiliser le descripteur de fichier inutilisé portant le numéro le plus bas, il utilise le numéro de descripteur **spécifié par l'utilisateur**.
+
+:::info
+Lors de l'appel de [dup2](https://docs.rs/nix/0.23.1/nix/unistd/fn.dup2.html) dans Rust, les paramètres donnés doivent être l'ancien descripteur de fichier et le nouveau.
+:::
+
+### Redirections
+Quand vous voulez rediriger stdout vers un fichier vous pouvez le faire en ouvrant un fichier à l'aide de l'appel système `open()`, puis en dupliquant ce descripteur de fichier sur `stdout` en utilisant `dup2()`, de sorte que la sortie standard soit maintenant représentée par le fichier ouvert et non par la console.
```
Comment (alexandruradovici) [inline docs/tp/04.md:48]: We need an example with `as_fd` instead of `open`.

**Review by alexandruradovici — CHANGES_REQUESTED** [review]

File: docs/tp/04.md [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+## Resources
+ 1. [OpenOptions](https://doc.rust-lang.org/std/fs/struct.OpenOptions.html)
+ 2. [Stdin](https://doc.rust-lang.org/std/io/fn.stdin.html)
+ 3. [Stdout](https://doc.rust-lang.org/std/io/fn.stdout.html)
+ 4. [dup](https://docs.rs/nix/0.23.1/nix/unistd/fn.dup.html)
+ 5. [dup2](https://docs.rs/nix/0.23.1/nix/unistd/fn.dup2.html)
+
+## Linux Standard Streams
+
+Sous Linux, `stdin` est le flux d'entrée standard. Cela accepte le texte comme entrée. La sortie texte de la commande au shell est transmise via le flux `stdout` (sortie standard). Les messages d'erreur de la commande sont envoyés via le flux `stderr` (erreur standard).
```
Comment (alexandruradovici) [inline docs/tp/04.md:17]: Both Windows and Linux use this, as `stdin` is part of the C standar library. Fd 0, 1 and 2 are Lunux specific.

**Review by alexandruradovici — CHANGES_REQUESTED** [review]

File: docs/tp/04.md (hunk 3 above)
Comment (Cristiana959) [reply docs/tp/04.md]: Done

File: docs/tp/04.md (hunk 4 above)
Comment (Cristiana959) [reply docs/tp/04.md]: Done

File: docs/tp/04.md [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+* 1: stdout
+* 2: stderr
+
+## Les appels système dup et dup2
+
+### dup
+L'appel système `dup()` crée **une copie** d'un descripteur de fichier.
+* Il utilise le descripteur inutilisé portant le numéro le plus bas pour le nouveau descripteur.
+* Si la copie est créée avec succès, les descripteurs de fichier d'origine et de copie peuvent être utilisés de manière interchangeable.
+* Ils se réfèrent tous les deux à la même description de fichier ouvert et partagent donc les indicateurs de décalage de fichier et d'état de fichier.
```
Comment (Ioan-Cristian) [inline docs/tp/04.md:38]:
```suggestion
* Ils se réfèrent tous les deux au même descripteur de fichier ouvert et partagent donc les indicateurs de décalage de fichier et d'état de fichier.
```

File: docs/tp/04.md [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+    println!("Hello, redirected stdout!");
+
+    Ok(())
+}
+```
+
+## Exercises
+1. Lisez votre nom à partir du clavier en utilisant [stdin](https://doc.rust-lang.org/std/io/fn.stdin.html). Imprimer le nom en utilisant [stdout](https://doc.rust-lang.org/std/io/fn.stdout.html). N'utilisez pas le println!
+2. Imprimer le contenu d'un fichier reçu en tant qu'argument de ligne de commande sur [stdout](https://doc.rust-lang.org/std/io/fn.stdout.html). N'utilisez pas le println!
+3. Lisez 3 chiffres sur le clavier en utilisant `stdin`. Écrivez le maximum de ces 3 nombres dans le fichier maximum.txt.
```
Comment (Ioan-Cristian) [inline docs/tp/04.md:77]:
```suggestion
3. Lisez 3 nombres depuis le clavier en utilisant `stdin`. Écrivez le maximum de ces 3 nombres dans le fichier maximum.txt.
```

File: docs/tp/04.md [hunk 7]
```diff
@@ -0,0 +1,84 @@
...
+4. Copiez le contenu d'un fichier donné comme argument depuis la ligne de commande dans un fichier nommé `copy.txt`.
+5. Ouvrez 2 instances du même fichier (ouvrez le fichier 2 fois). Lisez 10 octets de chaque instance et imprimez-les. Ouvrez une nouvelle instance du fichier et utilisez `dup()` pour faire une copie du descripteur de fichier de cette nouvelle instance. Lisez à partir des deux descripteurs de fichiers et voyez les différences entre la première situation et celle-ci.
+6. Écrivez la date d'aujourd'hui par `redirection` de stdout dans un fichier appelé `output.txt` en utilisant la fonction [write_all](https://doc.rust-lang.org/std/io/fn.stdout.html).
+7. Lisez 5 noms de filles à partir du clavier en utilisant [stdin](https://doc.rust-lang.org/std/io/fn.stdin.html) et ecrivez-les dans un fichier appelee `names.txt` en utilisant stdout. Executez encore une fois le programme et Lisez 5 noms de garçons à partir du clavier. Il faut **ajouter** les noms de garçons aux noms des filles deja existantes dans names.txt, pas d'écraser le fichier!
```
Comment (Ioan-Cristian) [inline docs/tp/04.md:81]:
```suggestion
7. Lisez 5 noms de filles à partir du clavier en utilisant [stdin](https://doc.rust-lang.org/std/io/fn.stdin.html) et ecrivez-les dans un fichier appelee `names.txt` en utilisant stdout. Executez encore une fois le programme et lisez 5 noms de garçons à partir du clavier. Il faut **ajouter** les noms de garçons aux noms des filles déjà existants dans names.txt, et non pas écraser le fichier! 
```

**Review by alexandruradovici — APPROVED** [review]

**Pull request merged**